File: packages/firestore/android/src/main/java/io/capawesome/capacitorjs/plugins/firebase/firestore/classes/options/WriteBatchOperation.java
package io.capawesome.capacitorjs.plugins.firebase.firestore.classes.options;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.JSObject;
import io.capawesome.capacitorjs.plugins.firebase.firestore.FirebaseFirestoreHelper;
import java.util.Map;
import org.json.JSONException;

public class WriteBatchOperation {

    private String type;
    private String reference;
    private Map<String, Object> data;

    @Nullable
    private SetOptions options;

    public WriteBatchOperation(@NonNull JSObject operation) throws JSONException {
        this.type = operation.getString("type");
        this.reference = operation.getString("reference");
        this.data = FirebaseFirestoreHelper.createHashMapFromJSONObject(operation.getJSObject("data"));

        if (operation.has("options") && operation.getJSObject("options") != null) {
            JSObject optsObj = operation.getJSObject("options");
            this.options = new SetOptions(optsObj);
        } else {
            this.options = null;
        }
    }

    public String getType() {
        return type;
    }

    public String getReference() {
        return reference;
    }

    public Map<String, Object> getData() {
        return data;
    }

    public SetOptions getOptions() {
        return options;
    }
}

File: packages/firestore/ios/Plugin/Classes/Options/WriteBatchOperation.swift
import Foundation
import Capacitor

@objc public class SetOptions: NSObject {
    private var merge: Bool

    init(_ options: JSObject) {
        self.merge = options["merge"] as? Bool ?? false
    }

    public func isMerge() -> Bool {
        return self.merge
    }
}

@objc public class WriteBatchOperation: NSObject {
    private var type: String
    private var reference: String
    private var data: [String: Any]
    private var options: SetOptions?

    init(_ operation: JSObject) {
        self.type = operation["type"] as? String ?? ""
        self.reference = operation["reference"] as? String ?? ""
        self.data = FirebaseFirestoreHelper.createHashMapFromJSObject(operation["data"] as? JSObject ?? [:])
        if let optsObj = operation["options"] as? JSObject {
            self.options = SetOptions(optsObj)
        } else {
            self.options = nil
        }
    }

    public func getType() -> String {
        return self.type
    }

    public func getReference() -> String {
        return self.reference
    }

    public func getData() -> [String: Any] {
        return self.data
    }

    public func getOptions() -> SetOptions? {
        return self.options
    }
}
